File: SenLib/Sen3/FileFixes/f2000_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class f2000_dat : FileMod {
		public string GetDescription() {
			return "Fix Hamilton gender (scene 2, text only).";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xbb3220cb9f85e554ul, 0xfd1196688bcc1ea5ul, 0x78a3d234u));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();
			var patcher = new SenScriptPatcher(bin);

			// switch gender in line from Angelica
			patcher.ExtendPartialCommand(0x46cf, 0x1b0, 0x482b, new byte[] { 0x73 });

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/f2000.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xbb3220cb9f85e554ul, 0xfd1196688bcc1ea5ul, 0x78a3d234u));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/f2000.dat", file) };
		}
	}
}

File: native/sen3/FileFixes/f2000_dat.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "../file_fixes.h"

#include "../../p3a/pack.h"
#include "../../p3a/structs.h"
#include "../../sen/sen_script_patcher.h"
#include "../../sha1.h"

namespace SenLib::Sen3::FileFixes::f2000_dat {
std::string_view GetDescription() {
    return "Fix Hamilton gender (scene 2, text only).";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_en/f2000.dat",
            88801,
            SenPatcher::SHA1FromHexString("bb3220cb9f85e554fd1196688bcc1ea578a3d234"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        SenScriptPatcher patcher(bin);

        // switch gender in line from Angelica
        patcher.ExtendPartialCommand(0x46cf, 0x1b0, 0x482b, {{0x73}});

        result.emplace_back(SenPatcher::P3APackFile{
            std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4});

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::f2000_dat
